Susi Crashes on Network State Changed

This Crash is occurring due to realm NoClassFound, on changing the Network state.
I/O Mobile data, Wi-fi.
Detected on:
Nexus 5 CM13 Android 6.0.1

![img_20161003_092010](https://cloud.githubusercontent.com/assets/7330961/19027127/f2c168cc-894a-11e6-87a6-4cded5a91bc9.jpg)

Log:
```
10-03 09:14:40.590 32622-32622/org.fossasia.susi.ai E/AndroidRuntime: FATAL EXCEPTION: main
                                                                      Process: org.fossasia.susi.ai, PID: 32622
                                                                      java.lang.RuntimeException: Unable to instantiate receiver io.realm.internal.network.NetworkStateReceiver: java.lang.ClassNotFoundException: Didn't find class "io.realm.internal.network.NetworkStateReceiver" on path: DexPathList[[zip file "/data/app/org.fossasia.susi.ai-1/base.apk"],nativeLibraryDirectories=[/data/app/org.fossasia.susi.ai-1/lib/arm, /data/app/org.fossasia.susi.ai-1/base.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]
                                                                          at android.app.ActivityThread.handleReceiver(ActivityThread.java:2750)
                                                                          at android.app.ActivityThread.-wrap14(ActivityThread.java)
                                                                          at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1440)
                                                                          at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                          at android.os.Looper.loop(Looper.java:148)
                                                                          at android.app.ActivityThread.main(ActivityThread.java:5466)
                                                                          at java.lang.reflect.Method.invoke(Native Method)
                                                                          at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                          at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                       Caused by: java.lang.ClassNotFoundException: Didn't find class "io.realm.internal.network.NetworkStateReceiver" on path: DexPathList[[zip file "/data/app/org.fossasia.susi.ai-1/base.apk"],nativeLibraryDirectories=[/data/app/org.fossasia.susi.ai-1/lib/arm, /data/app/org.fossasia.susi.ai-1/base.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]
                                                                          at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
                                                                          at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
                                                                          at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
                                                                          at android.app.ActivityThread.handleReceiver(ActivityThread.java:2745)
                                                                          at android.app.ActivityThread.-wrap14(ActivityThread.java) 
                                                                          at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1440) 
                                                                          at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                          at android.os.Looper.loop(Looper.java:148) 
                                                                          at android.app.ActivityThread.main(ActivityThread.java:5466) 
                                                                          at java.lang.reflect.Method.invoke(Native Method) 
                                                                          at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                          at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
                                                                      	Suppressed: java.lang.ClassNotFoundException: Didn't find class "io.realm.internal.network.NetworkStateReceiver" on path: DexPathList[[dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-support-annotations-24.2.1_2f4e12f50b74a0e56083d683ebcd01098f8544c5-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_9-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_8-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_7-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_6-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_5-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_4-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_3-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_2-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_1-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-slice_0-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-retrofit-2.1.0_cf279e13c7581267ec06efb423edffc451b4f5ff-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-realm-annotations-2.0.0_07f36a1f91e6acef3721afd25e88ec47ccc4492f-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-okio-1.8.0_6896106ae3c04439e55501a196384a5ecf60b5b1-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-okhttp-3.3.0_068abd747c7b414410f82e22c0d0b6acbd41650d-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-me.himanshusoni.chatmessageview-chat-message-view-1.0.3_7c27264fb08c446ab7c169a6481f8471c542675d-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-io.realm-realm-android-library-2.0.0_62fb1293c2ee6ddc69320de50e6dbecb6871db03-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-internal_impl-24.2.1_f360842d212d25afcf6d1d16c81608f47758e001-classes.dex", dex file "/data/data/org.fossasia.susi.ai/files/instant-run/dex/slice-internal_impl-24.2.1_ee35e25679ede1b1321
```

--- a/app/src/main/java/io/realm/internal/network/NetworkStateReceiver.java
+++ b/app/src/main/java/io/realm/internal/network/NetworkStateReceiver.java
@@ -1,0 +1,87 @@
+/**
+ * //TODO Fake Class, remove this class when realm 2.0.1 comes out.
+ * Copyright 2016 Realm Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.network;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import io.realm.internal.Util;
+
+/**
+ * This class is responsible for keeping track of system events related to the network so it can delegate them to
+ * interested parties.
+ */
+public class NetworkStateReceiver extends BroadcastReceiver {
+
+    private static List<ConnectionListener> listeners = new CopyOnWriteArrayList<ConnectionListener>();
+
+    /**
+     * Add a listener to be notified about any network changes.
+     * This method is thread safe.
+     * <p>
+     * IMPORTANT: Not removing it again will result in major leaks.
+     *
+     * @param listener the listener.
+     */
+    public static void addListener(ConnectionListener listener) {
+        listeners.add(listener);
+    }
+
+    /**
+     * Removes a network listener.
+     * This method is thread safe.
+     *
+     * @param listener the listener.
+     */
+    public static synchronized void removeListener(ConnectionListener listener) {
+        listeners.remove(listener);
+    }
+
+    /**
+     * Attempt to detect if a device is online and can transmit or receive data.
+     * This method is thread safe.
+     * <p>
+     * An emulator is always considered online, as `getActiveNetworkInfo()` does not report the correct value.
+     *
+     * @param context an Android context.
+     * @return {@code true} if device is online, otherwise {@code false}.
+     */
+    public static boolean isOnline(Context context) {
+        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = cm.getActiveNetworkInfo();
+        return ((networkInfo != null && networkInfo.isConnectedOrConnecting()) || Util.isEmulator());
+    }
+
+
+    public void onReceive(Context context, Intent intent) {
+        boolean connected = isOnline(context);
+        for (ConnectionListener listener : listeners) {
+            listener.onChange(connected);
+        }
+    }
+
+    public interface ConnectionListener {
+        void onChange(boolean connectionAvailable);
+    }
+}

--- a/app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java
+++ b/app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java
@@ -54,13 +54,12 @@ public class MainActivity extends AppCompatActivity {
     @BindView(R.id.send_message_layout)
     LinearLayout sendMessageLayout;
     RealmResults<ChatMessage> chatMessageDatabaseList;
-    private ChatFeedRecyclerAdapter recyclerAdapter;
-    private Realm realm;
-
     /**
      *  Preference for using Enter Key as send
      */
     SharedPreferences Enter_pref;
+    private ChatFeedRecyclerAdapter recyclerAdapter;
+    private Realm realm;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
@@ -141,9 +140,7 @@ public class MainActivity extends AppCompatActivity {
     }
 
     private void sendMessage(String query) {
-        ChatMessage chatMessage = new ChatMessage(query, true, false);
         updateDatabase(query, true, false);
-        recyclerAdapter.addMessage(chatMessage, true);
         computeOtherMessage(query);
     }
 
@@ -180,9 +177,7 @@ public class MainActivity extends AppCompatActivity {
     }
 
     private void addNewMessage(String answer) {
-        ChatMessage chatMessage = new ChatMessage(answer, false, false);
         updateDatabase(answer, false, false);
-        recyclerAdapter.addMessage(chatMessage, true);
     }
 
     private void updateDatabase(final String message, final boolean mine, final boolean image) {

--- a/app/src/main/java/org/fossasia/susi/ai/adapters/recyclerAdapters/ChatFeedRecyclerAdapter.java
+++ b/app/src/main/java/org/fossasia/susi/ai/adapters/recyclerAdapters/ChatFeedRecyclerAdapter.java
@@ -38,6 +38,15 @@ public class ChatFeedRecyclerAdapter extends RecyclerView.Adapter<ChatViewHolder
         this.itemList = itemList;
         this.currContext = curr_context;
         this.activity = activity;
+        itemList.addChangeListener(new RealmChangeListener<RealmResults<ChatMessage>>() {
+            @Override
+            public void onChange(RealmResults<ChatMessage> element) {
+                notifyItemInserted(ChatFeedRecyclerAdapter.this.itemList.size() - 1);
+                if (recyclerView != null) {
+                    recyclerView.smoothScrollToPosition(ChatFeedRecyclerAdapter.this.itemList.size() - 1);
+                }
+            }
+        });
     }
 
     @Override
@@ -52,24 +61,6 @@ public class ChatFeedRecyclerAdapter extends RecyclerView.Adapter<ChatViewHolder
         this.recyclerView = null;
     }
 
-    public void addMessage(final ChatMessage chatMessage, final boolean shouldScrollToBottom) {
-        if (itemList == null) {
-        }
-        itemList.addChangeListener(new RealmChangeListener<RealmResults<ChatMessage>>() {
-            @Override
-            public void onChange(RealmResults<ChatMessage> element) {
-                notifyItemInserted(itemList.size() - 1);
-                if (recyclerView != null && shouldScrollToBottom) {
-                    recyclerView.smoothScrollToPosition(itemList.size() - 1);
-                }
-            }
-        });
-
-    }
-
-    public void addMessage(ChatMessage chatMessage) {
-        addMessage(chatMessage, false);
-    }
 
     @Override
     public ChatViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {

